### ISSUE TITLE
systemd spams with Time has been changed messages

### ISSUE THREAD

**Issue by kimmoli** [kimmoli/sfos-onyx-issues#14]
```
May 22 12:10:25 localhost systemd[1033]: Time has been changed
May 22 12:10:28 localhost systemd[1]: Time has been changed
May 22 12:10:28 localhost systemd[1033]: Time has been changed
May 22 12:10:30 localhost systemd[1]: Time has been changed
May 22 12:10:30 localhost systemd[1033]: Time has been changed
May 22 12:10:32 localhost systemd[1]: Time has been changed
May 22 12:10:32 localhost systemd[1033]: Time has been changed
May 22 12:10:58 localhost systemd[1]: Time has been changed
May 22 12:10:58 localhost systemd[1033]: Time has been changed
May 22 12:11:02 localhost systemd[1]: Time has been changed
May 22 12:11:02 localhost systemd[1033]: Time has been changed
May 22 12:11:03 localhost systemd[1]: Time has been changed
May 22 12:11:03 localhost systemd[1033]: Time has been changed
May 22 12:11:17 localhost systemd[1]: Time has been changed
May 22 12:11:17 localhost systemd[1033]: Time has been changed
May 22 12:11:28 localhost systemd[1]: Time has been changed
May 22 12:11:28 localhost systemd[1033]: Time has been changed
May 22 12:11:29 localhost systemd[1]: Time has been changed
May 22 12:11:29 localhost systemd[1033]: Time has been changed
May 22 12:11:30 localhost systemd[1]: Time has been changed
May 22 12:11:30 localhost systemd[1033]: Time has been changed
May 22 12:11:32 localhost systemd[1]: Time has been changed
May 22 12:11:32 localhost systemd[1033]: Time has been changed
```

**Comment by kimmoli** [comment]
https://git.merproject.org/kimmoli/systemd/commit/2b65db3880687e214386127f27f731233a978413

**Comment by kimmoli** [comment]
@taaem what about this one? i hate spinning own systemd fork, but it makes debugging much easier imo

**Comment by taaem** [comment]
Why not make a PR to upstream Mer systemd? I mean that should be a problem for them too

**Comment by kimmoli** [comment]
I haven't seen this on sbj, C, tablet. So there must be a root-cause for it - this just hides the loggings.

**Comment by izh1979** [comment]
Systemd argues about system time change. It could happen either because of some daemon or after waking up. I recall something related to it...

**Comment by kimmoli** [comment]
There is a simple scenario how i can repeat this flood.
- ssh into the phone, and `devel-su journalctl -fa`
- open another ssh session to the phone
- blank screen and wait few secs to the phone to sleep
- type in the another session
- journal will print _time has been changed_ on every keypress

so waking up might be the cause (?)

**Comment by kimmoli** [comment]
Actually (after re-checking) JollaC spams a lot more in similar situation, but already the journal prints seems to self-feed this...

```
Jul 06 14:37:04 jollaC systemd[1]: Time has been changed
Jul 06 14:37:04 jollaC systemd[670]: Time has been changed
Jul 06 14:37:04 jollaC kernel: Freezing user space processes ... (elapsed 0.004 seconds) done.
Jul 06 14:37:04 jollaC kernel: Freezing remaining freezable tasks ... (elapsed 0.002 seconds) done.
Jul 06 14:37:04 jollaC kernel: Suspending console(s) (use no_console_suspend to debug)
Jul 06 14:37:04 jollaC kernel: PM: suspend of devices complete after 41.093 msecs
Jul 06 14:37:04 jollaC kernel: PM: late suspend of devices complete after 2.985 msecs
Jul 06 14:37:04 jollaC kernel: PM: noirq suspend of devices complete after 3.296 msecs
Jul 06 14:37:04 jollaC kernel: Disabling non-boot CPUs ...
Jul 06 14:37:04 jollaC kernel: CPU0:msm_cpu_pm_enter_sleep mode:3 during suspend
Jul 06 14:37:04 jollaC kernel: Enabling non-boot CPUs ...
Jul 06 14:37:04 jollaC kernel: CPU1 is up
Jul 06 14:37:04 jollaC kernel: CPU2 is up
Jul 06 14:37:04 jollaC kernel: CPU3 is up
Jul 06 14:37:04 jollaC kernel: PM: noirq resume of devices complete after 1.551 msecs
Jul 06 14:37:04 jollaC kernel: PM: early resume of devices complete after 2.181 msecs
Jul 06 14:37:04 jollaC kernel: PM: resume of devices complete after 94.887 msecs
Jul 06 14:37:04 jollaC kernel: Restarting tasks ... done.
Jul 06 14:37:04 jollaC kernel: PM: suspend exit 2016-07-06 11:37:04.370860532 UTC
Jul 06 14:37:04 jollaC kernel: PM: suspend entry 2016-07-06 11:37:04.872134959 UTC
Jul 06 14:37:05 jollaC kernel: PM: Syncing filesystems ... done.
Jul 06 14:37:05 jollaC kernel: Freezing user space processes ... (elapsed 0.003 seconds) done.
Jul 06 14:37:05 jollaC systemd[1]: Time has been changed
Jul 06 14:37:05 jollaC systemd[670]: Time has been changed
....
```

**Comment by izh1979** [comment]
https://lists.freedesktop.org/archives/systemd-commits/2012-November/002857.html
It seems it about this functionality that waits for kernel notifications about CLOCK_REALTIME jumps. The question is whether the kernel should notify apps this way after each wake-up.

**Comment by izh1979** [comment]
I see similar logic in new systemd versions too. So the question is whether the kernel works properly with it.

**Comment by izh1979** [comment]
"The timerfd_settime() system call adds a TFD_TIMER_CANCEL_ON_SET flag. If this flag is set for a CLOCK_REALTIME absolute (TFD_TIMER_ABSTIME) timer, then the timer is expired if the clock is reset."

So I would dig into the kernel looking for TFD_TIMER_CANCEL_ON_SET handling. I'm not sure that it should be triggered on a wake-up since the clocks were adjusted by the kernel itself (if I understand correctly) and not by user-space (like NTP daemon).

**Comment by izh1979** [comment]
It could be that this part is SoC-dependent.

**Comment by kimmoli** [comment]
http://lwn.net/Articles/429925/ - Android Alarm Timers
and
https://github.com/systemd/systemd/blob/master/src/core/manager.c#L361
?

So shutting up this from LOG_INFO could be a reasonable solution not affecting the actual functionality behind?

**Comment by izh1979** [comment]
Well, disabling this message could help.
Although I'm not sure about:

1) Whether system clock adjustment notifications should be send to userland
    on each wake-up.

2) If 1) is normal, shouldn't systemd use some other clock source for such
    events.

It could be that the systemd is not well tested on mobile/android kernels
because it was primarily designed for PCs/servers and not so many mobile
OSes (probably only SFOS and Tizen) use is as init daemon.

**Comment by spiiroin** [comment]
@izh1979  having (1) is one of the few ways available for userland to know that suspend has occurred, so I would not change kernel side behavior.

If we want to retain logging about actual system time changes, we could filter out suspend-resume cycles at systemd side by checking if delta between monotime and boot time has changed (i.e. log realtime changes only if monotime-boottime does not change).

**Comment by kimmoli** [comment]
This time change is triggered in another place in systemd too
https://github.com/systemd/systemd/blob/5f36e3d30375cf04292bbc1bf3f4d7512cf80139/src/timesync/timesyncd-manager.c#L286
which "skips its own jumps" whatever that means
https://github.com/systemd/systemd/blob/5f36e3d30375cf04292bbc1bf3f4d7512cf80139/src/timesync/timesyncd-manager.c#L273

**Comment by izh1979** [comment]
Systemd now contains time adjusting code: https://github.com/systemd/systemd/blob/5f36e3d30375cf04292bbc1bf3f4d7512cf80139/src/timesync/timesyncd-manager.c#L347
So yes, it can change time by itself.

**Comment by kimmoli** [comment]
ah, sorry. sure i looked at HEAD, not the stick&stone version

**Comment by kimmoli** [comment]
yeah, clock_was_set comes from hrtimers_resume()

https://github.com/kimmoli/android_kernel_oneplus_msm8974/blob/issue14/kernel/hrtimer.c#L750

dmesg
```
[20161211.000000] 20161211_19:06:54.528091]@0 kimmo was here
[20161211.000000] 20161211_19:06:54.225149]@0 kimmo was here
[20161211.000000] 20161211_19:06:55.654806]@0 kimmo was here
[20161211.000000] 20161211_19:06:55.055319]@0 kimmo was here
[20161211.000000] 20161211_19:06:56.463708]@0 kimmo was here
[20161211.000000] 20161211_19:06:56.824542]@0 kimmo was here
[20161211.000000] 20161211_19:06:58.195620]@0 kimmo was here
[20161211.000000] 20161211_19:07:44.521674]@0 kimmo was here
[20161211.000000] 20161211_19:07:59.092271]@0 kimmo was here
[20161211.000000] 20161211_19:08:02.456692]@0 kimmo was here
[20161211.000000] 20161211_19:08:04.163251]@0 kimmo was here
```
journal
```
Dec 11 13:06:54 Sailfish systemd[1015]: Time has been changed
Dec 11 13:06:54 Sailfish systemd[1015]: Time has been changed
Dec 11 13:06:55 Sailfish systemd[1015]: Time has been changed
Dec 11 13:06:55 Sailfish systemd[1015]: Time has been changed
Dec 11 13:06:56 Sailfish systemd[1015]: Time has been changed
Dec 11 13:06:56 Sailfish systemd[1015]: Time has been changed
Dec 11 13:06:58 Sailfish systemd[1015]: Time has been changed
Dec 11 13:07:44 Sailfish systemd[1015]: Time has been changed
Dec 11 13:07:59 Sailfish systemd[1015]: Time has been changed
Dec 11 13:08:02 Sailfish systemd[1015]: Time has been changed
Dec 11 13:08:04 Sailfish systemd[1015]: Time has been changed
```
(Wondering also why kernel timestamp is +4 hours?)

**Comment by izh1979** [comment]
Is it in UTC?

**Comment by kimmoli** [comment]
FI (EET) is UTC+2 during winter, so i don't think so

**Comment by kimmoli** [comment]
I have no idea what i'm doing...

https://git.merproject.org/kimmoli/systemd/commit/23c191e16b5192427600b4779d9855fa2b3e8513

Maybe need more resolution? in 1-sec resolution, those seems to be same at time of printing the log...

but anyways, when setting time with `date 12111100` it seems to trigger also DSME.

```
Dec 11 14:37:39 Sailfish systemd[1168]: Time has been changed boot Sun Dec 11 14:37:39 current Sun Dec 11 14:37:39 real Sun Dec 11 14:37:39
Dec 11 14:37:39 Sailfish systemd[1]: Time has been changed boot Sun Dec 11 14:37:39 current Sun Dec 11 14:37:39 real Sun Dec 11 14:37:39
Dec 11 14:37:39 Sailfish systemd[1168]: Time has been changed boot Sun Dec 11 14:37:39 current Sun Dec 11 14:37:39 real Sun Dec 11 14:37:39
Dec 11 14:37:44 Sailfish systemd[1]: Time has been changed boot Sun Dec 11 14:37:44 current Sun Dec 11 14:37:44 real Sun Dec 11 14:37:44
Dec 11 14:37:44 Sailfish systemd[1168]: Time has been changed boot Sun Dec 11 14:37:44 current Sun Dec 11 14:37:44 real Sun Dec 11 14:37:44
Dec 11 11:00:00 Sailfish DSME[773]: IPHB: rtc delta to 1448405584
Dec 11 11:00:00 Sailfish systemd[1168]: Time has been changed boot Sun Dec 11 11:00:00 current Sun Dec 11 11:00:00 real Sun Dec 11 11:00:00
Dec 11 11:00:00 Sailfish systemd[1]: Time has been changed boot Sun Dec 11 11:00:00 current Sun Dec 11 11:00:00 real Sun Dec 11 11:00:00
Dec 11 11:00:00 Sailfish systemd[1]: Time has been changed boot Sun Dec 11 11:00:00 current Sun Dec 11 11:00:00 real Sun Dec 11 11:00:00
Dec 11 11:00:00 Sailfish systemd[1168]: Time has been changed boot Sun Dec 11 11:00:00 current Sun Dec 11 11:00:00 real Sun Dec 11 11:00:00

```

**Comment by izh1979** [comment]
By the way, it's interesting why DSME doesn't spam too? It it using different method of detecting time changes?

**Comment by spiiroin** [comment]
@izh1979: resume -> realtime change notification -> timed wakes up & sends signal -> dsme wakes up & does what it needs to do = not much as the realtime vs rtc diff stays the same as expected -> from dsme pov there is nothing too interesting going on, so there is/should be no need to go on logging spree.

@kimmoli: date change -> realtime change ... -> dsme wakes up and notices that realtime vs rtc diff *has* changed -> action taken and logged (probably even that could be demoted so that it does not show up in default verbosity)

**Comment by kimmoli** [comment]
https://github.com/systemd/systemd/issues/4896

**Comment by rinigus** [comment]
according to @kimmoli , fixed in 2.1.4.14
